### PR TITLE
Initial push of schreefstock functions

### PR DESCRIPTION
**Pull request by noideas4thisname**
Connection string for Azure already in Heroku.

~~Ugly~~ Getting better every day but functional.

Currently only supports:
- Checking balance
- Resetting to starting value
- Seeing all balances
- Sending to other users

### CODE REVIEW

File: handlers/stock.js [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+module.exports = {
+    name: 'stock',
+    command: 'stock',
+    handler: stock
+};
+
+function stock(channel, message, params) {
+
+    // Wont work until this is populated from whatever is managing creds
+    var placeholderConnectionString;
```
Comment (skeltonath) [inline handlers/stock.js]: for local testing make a `.env` file and put env vars in it

for production, we need to add those vars to the heroku node

File: handlers/stock.js [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+    // Initialize table service
+    const tableSvc = azure.createTableService(placeholderConnectionString);
+    var splitParams = params.split(" ");
```
Comment (skeltonath) [inline handlers/stock.js]: for ECMA6 use `const` for unchanging variables and `let` for others

File: handlers/stock.js [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+    // Verify the Azure Storage Table exists and create if it doesn't
+    tableSvc.createTableIfNotExists(TABLENAME, function (error, result, response) {
+        if (!error) {
```
Comment (skeltonath) [inline handlers/stock.js]: log the error if it isn't null maybe?

File: handlers/stock.js [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+            // Table exists or created
+        }
+    })
+
+    // Verify all users have an account and create new ones for those that don't
+    for (let member of channel.guild.members.array()) {
```
Comment (skeltonath) [inline handlers/stock.js]: `channel.guild.members.forEach(member => {
  ...
});
`

File: handlers/stock.js [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+        getStock(tableSvc, member.user.username, function (callback) {
+            if (typeof callback != 'number') {
```
Comment (skeltonath) [inline handlers/stock.js]: when comparing numbers and strings in javascript use `===` or `!==`

File: handlers/stock.js [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+                console.log("initializing " + member.user.username);
```
Comment (skeltonath) [inline handlers/stock.js]: for logging stuff add this to the top of the file
```
const log4js  = require('log4js');
const LOG = log4js.getLogger('schreefstock');
```

then use LOG like
`LOG.info('foo');`

File: handlers/stock.js [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+                updateUser(tableSvc, member.user.username, STARTINGSTOCK);
+            }
+        })
+    }
+
+    if (splitParams[0] == "help") {
```
Comment (skeltonath) [inline handlers/stock.js]: for strings use single quotes `'` whenever possible

File: handlers/stock.js [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+                if (typeof receiverStock == 'number') {
+                    getStock(tableSvc, message.author.username, function (senderStock) {
+                        if ((senderStock - sendAmount) > 0) {
+
+                            var newSenderTotal = senderStock - sendAmount;
+                            var newReceiverTotal = receiverStock + sendAmount;
+
+                            console.log("Attempting to send " + sendAmount + " stock");
+
+                            updateUser(tableSvc, message.author.username, newSenderTotal);
```
Comment (skeltonath) [inline handlers/stock.js]: any worry about race conditions or rollbacks?

don't want people creating money from nowhere

File: handlers/stock.js [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+                                message.author.username + ': ' + senderStock + ' - ' + sendAmount + " = " + newSenderTotal + "\n" +
+                                recipient + ': ' + receiverStock + ' + ' + sendAmount + " = " + newReceiverTotal);
+                        }
+                        else {
+                            channel.send("Not enough stock to make the transaction.")
+                        }
+                    })
+                }
+                else {
+                    channel.send("That user doesn't exist! Shcreef doesn't throw his stock to the abyss.")
```
Comment (skeltonath) [inline handlers/stock.js]: misspelled 'schreef'

File: handlers/stock.js [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+
+    var userEntry = {
+        PartitionKey: ENTGEN.String(username),
+        RowKey: ENTGEN.String('1'),
+        schreefstock: ENTGEN.Int32(amount),
+    }
+
+    tableSvc.insertOrReplaceEntity(TABLENAME, userEntry, null, function (error, result, response) {
+        if (!error) {
+            // Entity updated
```
Comment (skeltonath) [inline handlers/stock.js]: log error

File: handlers/stock.js [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+
+    // Reset command to go back to starting stock
+    if (splitParams[0] == "reset") {
+        updateUser(tableSvc, message.author.username, STARTINGSTOCK);
+    }
+
+    // Get the stock for that user
+    if (splitParams[0] == "balance") {
+        getStock(tableSvc, message.author.username, function (callback) {
+            channel.send(message.author.username + "'s balance is " + callback + "$$");
```
Comment (skeltonath) [inline handlers/stock.js]: javascript now allows you to insert variables right into strings
```
`string text ${expression} string text`
```

**Comment by skeltonath** [comment]
also get rid of that package-lock.json

and fix those conflicts with a rebase + yarn install

File: handlers/stock.js (hunk 8 above)
Comment (noideas4thisname) [reply handlers/stock.js]: I'll look into this more later, but the basic logic:
- if getStock fails for either the recipient or sender, neither receiver or sender would be updated
- the sender can't go below 0 stock by sending more than they have

I thought the callbacks protected it from a race but let me know if there's something I'm missing

**Comment by noideas4thisname** [comment]
Most things should be fixed. Branch history is a mess, but take another look.

File: handlers/stock.js [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+                            transactionString += `${message.author.username}: ${senderStock} - ${sendAmount} = ${newSenderTotal}\n
+                                                  ${recipient}: ${receiverStock} + ${sendAmount} = ${newReceiverTotal}`
+
+                            LOG.info(transactionString);
+                            channel.send(transactionString);
+                        }
+                        else {
+                            channel.send('Not enough stock to make the transaction.')
+                        }
+                    })
```
Comment (skeltonath) [inline handlers/stock.js]: missing ;

File: handlers/stock.js [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+                }
+                else {
+                    channel.send('That user does not exist! Schreef won\'t throw his stock to the abyss.');
+                }
+            })
```
Comment (skeltonath) [inline handlers/stock.js]: missing ;

File: handlers/stock.js [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+
+// Displays everyone's stock
+function showAllStock(tableSvc, channel) {
+    const query = new azure.TableQuery().where('RowKey eq ?', '1');
+    let displayResult = 'Full ledger: ';
+
+    tableSvc.queryEntities(TABLENAME, query, null, function (error, result, response) {
+        if (!error) {
+            // query successful
+            for (let i = 0, len = result.entries.length; i < len; i++) {
```
Comment (skeltonath) [inline handlers/stock.js]: can you not use forEach or map on result.entries.length?

File: handlers/stock.js [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+    }
+
+    tableSvc.insertOrReplaceEntity(TABLENAME, userEntry, null, function (error, result, response) {
+        if (!error) {
+            // Entity updated
+        }
+        else {
+            LOG.error(error);
+        }
+    })
```
Comment (skeltonath) [inline handlers/stock.js]: missing ;

File: handlers/stock.js [hunk 16]
```diff
@@ -0,0 +1,176 @@
...
+                displayResult = `${displayResult}\n${line}`;
+            }
+            displayResult = `${displayResult}\nschreef has an infinite amount of stock`;
+
+            channel.send(displayResult);
+        }
+        else {
+            LOG.error(error);
+        }
+    })
```
Comment (skeltonath) [inline handlers/stock.js]: missing ;

File: handlers/stock.js [hunk 17]
```diff
@@ -0,0 +1,176 @@
...
+    const query = new azure.TableQuery().select(['schreefstock']).where('PartitionKey eq ?', username);
+
+    tableSvc.queryEntities(TABLENAME, query, null, function (error, result, response) {
+        if (!error) {
+            // query successful
+            if (result.entries.length > 0) {
+                callback(result.entries[0].schreefstock._);
+            }
+            else {
+                callback(null);
```
Comment (skeltonath) [inline handlers/stock.js]: typically in a callback, in a failure, you pass both `null` and `error` so the callback can handle appropriately

File: handlers/stock.js (hunk 14 above)
Comment (noideas4thisname) [reply handlers/stock.js]: probably can - i'll rewrite it later today

File: handlers/stock.js (hunk 12 above)
Comment (noideas4thisname) [reply handlers/stock.js]: for all of these comments -these arent syntactically necessary right? This is just for cleanliness / consistency?

File: handlers/stock.js (hunk 17 above)
Comment (noideas4thisname) [reply handlers/stock.js]: i'll add these

File: handlers/stock.js (hunk 12 above)
Comment (skeltonath) [reply handlers/stock.js]: if you only have a single statement in a block, like you do here, i think it works fine

but it is _technically_ incorrect to not have semicolons

i'm gonna get a eslint/jshint thing setup to let people know about of these issues on build/run, so we can keep our style consistent

File: handlers/stock.js (hunk 7 above)
Comment (stffrd) [reply handlers/stock.js]: I disagree with single quotes over double quotes. 

But also this is an issue that'll just fix itself when ESLINT is thrown down anyway.

File: handlers/stock.js (hunk 1 above)
Comment (noideas4thisname) [reply handlers/stock.js]: I read some back and forth articles on it and bow to whatever style guide we choose. Converted fully to single at the moment and leaving it that way until for now.

File: handlers/stock.js (hunk 7 above)
Comment (noideas4thisname) [reply handlers/stock.js]: Moving comment as fasthub put it in a weird spot:

I read some back and forth articles on it and bow to whatever style guide we choose. Converted fully to single at the moment and leaving it that way for now.

File: handlers/stock.js (hunk 10 above)
Comment (noideas4thisname) [reply handlers/stock.js]: done in else case now

**Comment by noideas4thisname** [comment]
Extra note here: This works for now and is ready to merge with others blessing. I'll submit a new PR later with some improvements to the logic to make it cleaner. Mainly:
- create an array of supported params
- getting rid of the chained "ifs" in favor of switching on params
- putting each .stock <param> into its own function

...and whatever else seems right when I have more time to revisit.

**Pull request merged**